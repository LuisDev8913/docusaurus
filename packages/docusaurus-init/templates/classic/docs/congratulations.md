---
title: Congratulations!
---

Congratulations on making it this far!

You have learned the **basics of Docusaurus** and made some changes to the **initial template**.

Docusaurus has **much more to offer**!

Have 5 more minutes? Take a look at **[versioning](./manage-docs-versions.md)** and **[i18n](./translate-your-site.md)**.

## What's next?

- Read the [official documentation](https://docusaurus.io/).
- Add a custom [Design and Layout](https://docusaurus.io/docs/styling-layout)
- Add a [search bar](https://docusaurus.io/docs/search)
- Find inspirations in the [Docusaurus showcase](https://docusaurus.io/showcase)
- Get involved in the [Docusaurus Community](https://docusaurus.io/community/support)
