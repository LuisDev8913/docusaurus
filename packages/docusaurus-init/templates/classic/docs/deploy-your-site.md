---
title: Deploy your site
---

Docusaurus is a **static-site-generator** (also called [Jamstack](https://jamstack.org/)), and builds your site as **static HTML, JavaScript and CSS files**.

## Build your site

Build your site **for production**:

```bash
npm run build
```

The static files are generated in the `build` directory.

## Deploy your site

Test your production build locally:

```bash
npm run serve
```

The `build` folder is now served at `http://localhost:3000/`.

You can now deploy the `build` folder **almost anywhere** easily, **for free** or very small cost (read the **[Deployment Guide](https://v2.docusaurus.io/docs/deployment)**).
