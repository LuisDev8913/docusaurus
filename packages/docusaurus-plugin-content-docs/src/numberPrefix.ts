/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const NumberPrefixRegex = /^(?<numberPrefix>\d+)(?<separator>\s*[-_.]+\s*)(?<suffix>.*)$/;

// 0-myDoc => myDoc
export function stripNumberPrefix(str: string) {
  return NumberPrefixRegex.exec(str)?.groups?.suffix ?? str;
}

// 0-myFolder/0-mySubfolder/0-myDoc => myFolder/mySubfolder/myDoc
export function stripPathNumberPrefixes(path: string) {
  return path.split('/').map(stripNumberPrefix).join('/');
}

// 0-myDoc => {filename: myDoc, numberPrefix: 0}
// 003 - myDoc => {filename: myDoc, numberPrefix: 3}
export function extractNumberPrefix(
  filename: string,
): {filename: string; numberPrefix?: number} {
  const match = NumberPrefixRegex.exec(filename);
  const cleanFileName = match?.groups?.suffix ?? filename;
  const numberPrefixString = match?.groups?.numberPrefix;
  const numberPrefix = numberPrefixString
    ? parseInt(numberPrefixString, 10)
    : undefined;
  return {
    filename: cleanFileName,
    numberPrefix,
  };
}
