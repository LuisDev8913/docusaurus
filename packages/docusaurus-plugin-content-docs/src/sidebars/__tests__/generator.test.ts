/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {DefaultSidebarItemsGenerator} from '../generator';
import type {Sidebar, SidebarItemsGenerator} from '../types';
import {DefaultNumberPrefixParser} from '../../numberPrefix';
import {isCategoryIndex} from '../../docs';

describe('DefaultSidebarItemsGenerator', () => {
  function testDefaultSidebarItemsGenerator(
    params: Partial<Parameters<SidebarItemsGenerator>[0]>,
  ) {
    return DefaultSidebarItemsGenerator({
      numberPrefixParser: DefaultNumberPrefixParser,
      isCategoryIndex,
      item: {
        type: 'autogenerated',
        dirName: '.',
      },
      version: {
        versionName: 'current',
        contentPath: 'docs',
      },
      docs: [],
      options: {
        sidebarCollapsed: true,
        sidebarCollapsible: true,
      },
      categoriesMetadata: {},
      ...params,
    });
  }

  test('generates empty sidebar slice when no docs and emit a warning', async () => {
    const consoleWarn = jest.spyOn(console, 'warn');
    const sidebarSlice = await testDefaultSidebarItemsGenerator({
      docs: [],
    });
    expect(sidebarSlice).toEqual([]);
    expect(consoleWarn).toHaveBeenCalledWith(
      expect.stringMatching(
        /.*\[WARNING\].* No docs found in .*\..*: can't auto-generate a sidebar\..*/,
      ),
    );
  });

  test('generates simple flat sidebar', async () => {
    const sidebarSlice = await DefaultSidebarItemsGenerator({
      numberPrefixParser: DefaultNumberPrefixParser,
      item: {
        type: 'autogenerated',
        dirName: '.',
      },
      version: {
        versionName: 'current',
        contentPath: '',
      },
      docs: [
        {
          id: 'doc1',
          source: 'doc1.md',
          sourceDirName: '.',
          sidebarPosition: 2,
          frontMatter: {
            sidebar_label: 'doc1 sidebar label',
          },
        },
        {
          id: 'doc2',
          source: 'doc2.md',
          sourceDirName: '.',
          sidebarPosition: 3,
          frontMatter: {},
        },
        {
          id: 'doc3',
          source: 'doc3.md',
          sourceDirName: '.',
          sidebarPosition: 1,
          frontMatter: {},
        },
        {
          id: 'doc4',
          source: 'doc4.md',
          sourceDirName: '.',
          sidebarPosition: 1.5,
          frontMatter: {},
        },
        {
          id: 'doc5',
          source: 'doc5.md',
          sourceDirName: '.',
          sidebarPosition: undefined,
          frontMatter: {},
        },
      ],
      options: {
        sidebarCollapsed: true,
        sidebarCollapsible: true,
      },
    });

    expect(sidebarSlice).toEqual([
      {type: 'doc', id: 'doc3'},
      {type: 'doc', id: 'doc4'},
      {type: 'doc', id: 'doc1', label: 'doc1 sidebar label'},
      {type: 'doc', id: 'doc2'},
      {type: 'doc', id: 'doc5'},
    ] as Sidebar);
  });

  test('generates complex nested sidebar', async () => {
    const sidebarSlice = await DefaultSidebarItemsGenerator({
      numberPrefixParser: DefaultNumberPrefixParser,
      isCategoryIndex,
      item: {
        type: 'autogenerated',
        dirName: '.',
      },
      version: {
        versionName: 'current',
        contentPath: '',
      },
      categoriesMetadata: {
        '02-Guides': {collapsed: false},
        '02-Guides/01-SubGuides': {
          label: 'SubGuides (metadata file label)',
          link: {
            type: 'generated-index',
            slug: 'subguides-generated-index-slug',
            title: 'subguides-title',
            description: 'subguides-description',
          },
        },
      },
      docs: [
        {
          id: 'intro',
          source: '@site/docs/intro.md',
          sourceDirName: '.',
          sidebarPosition: 1,
          frontMatter: {},
        },
        {
          id: 'tutorials-index',
          source: '@site/docs/01-Tutorials/index.md',
          sourceDirName: '01-Tutorials',
          sidebarPosition: 2,
          frontMatter: {},
        },
        {
          id: 'tutorial2',
          source: '@site/docs/01-Tutorials/tutorial2.md',
          sourceDirName: '01-Tutorials',
          sidebarPosition: 2,
          frontMatter: {},
        },
        {
          id: 'tutorial1',
          source: '@site/docs/01-Tutorials/tutorial1.md',
          sourceDirName: '01-Tutorials',
          sidebarPosition: 1,
          frontMatter: {},
        },
        {
          id: 'guides-index',
          source: '@site/docs/02-Guides/02-Guides.md', // TODO should we allow to just use "Guides.md" to have an index?
          sourceDirName: '02-Guides',
          frontMatter: {},
        },
        {
          id: 'guide2',
          source: '@site/docs/02-Guides/guide2.md',
          sourceDirName: '02-Guides',
          sidebarPosition: 2,
          frontMatter: {},
        },
        {
          id: 'guide1',
          source: '@site/docs/02-Guides/guide1.md',
          sourceDirName: '02-Guides',
          sidebarPosition: 1,
          frontMatter: {
            sidebar_class_name: 'foo',
          },
        },
        {
          id: 'nested-guide',
          source: '@site/docs/02-Guides/01-SubGuides/nested-guide.md',
          sourceDirName: '02-Guides/01-SubGuides',
          sidebarPosition: undefined,
          frontMatter: {},
        },
        {
          id: 'end',
          source: '@site/docs/end.md',
          sourceDirName: '.',
          sidebarPosition: 3,
          frontMatter: {},
        },
      ],
      options: {
        sidebarCollapsed: true,
        sidebarCollapsible: true,
      },
    });

    expect(sidebarSlice).toEqual([
      {type: 'doc', id: 'intro'},
      {
        type: 'category',
        label: 'Tutorials',
        link: {
          type: 'doc',
          id: 'tutorials-index',
        },
        items: [
          {type: 'doc', id: 'tutorial1'},
          {type: 'doc', id: 'tutorial2'},
        ],
      },
      {
        type: 'category',
        label: 'Guides',
        link: {
          type: 'doc',
          id: 'guides-index',
        },
        collapsed: false,
        items: [
          {type: 'doc', id: 'guide1', className: 'foo'},
          {
            type: 'category',
            label: 'SubGuides (metadata file label)',
            items: [{type: 'doc', id: 'nested-guide'}],
            link: {
              type: 'generated-index',
              slug: 'subguides-generated-index-slug',
              title: 'subguides-title',
              description: 'subguides-description',
            },
          },
          {type: 'doc', id: 'guide2'},
        ],
      },
      {type: 'doc', id: 'end'},
    ] as Sidebar);
  });

  test('generates subfolder sidebar', async () => {
    // Ensure that category metadata file is correctly read
    // fix edge case found in https://github.com/facebook/docusaurus/issues/4638
    const sidebarSlice = await DefaultSidebarItemsGenerator({
      numberPrefixParser: DefaultNumberPrefixParser,
      isCategoryIndex,
      item: {
        type: 'autogenerated',
        dirName: 'subfolder/subsubfolder',
      },
      version: {
        versionName: 'current',
        contentPath: '',
      },
      categoriesMetadata: {
        'subfolder/subsubfolder/subsubsubfolder2': {
          position: 2,
          label: 'subsubsubfolder2 (_category_.yml label)',
          className: 'bar',
        },
        'subfolder/subsubfolder/subsubsubfolder3': {
          position: 1,
          label: 'subsubsubfolder3 (_category_.json label)',
          link: {
            type: 'doc',
            id: 'doc1', // This is a "fully-qualified" ID that can't be found locally
          },
        },
      },
      docs: [
        {
          id: 'doc1',
          source: 'doc1.md',
          sourceDirName: 'subfolder/subsubfolder',
          sidebarPosition: undefined,
          frontMatter: {},
        },
        {
          id: 'doc2',
          source: 'doc2.md',
          sourceDirName: 'subfolder',
          sidebarPosition: undefined,
          frontMatter: {},
        },
        {
          id: 'doc3',
          source: 'doc3.md',
          sourceDirName: '.',
          sidebarPosition: undefined,
          frontMatter: {},
        },
        {
          id: 'doc4',
          source: 'doc4.md',
          sourceDirName: 'subfolder/subsubfolder',
          sidebarPosition: undefined,
          frontMatter: {},
        },
        {
          id: 'doc5',
          source: 'doc5.md',
          sourceDirName: 'subfolder/subsubfolder/subsubsubfolder',
          sidebarPosition: undefined,
          frontMatter: {},
        },
        {
          id: 'doc6',
          source: 'doc6.md',
          sourceDirName: 'subfolder/subsubfolder/subsubsubfolder2',
          sidebarPosition: undefined,
          frontMatter: {},
        },
        {
          id: 'doc7',
          source: 'doc7.md',
          sourceDirName: 'subfolder/subsubfolder/subsubsubfolder3',
          sidebarPosition: 2,
          frontMatter: {},
        },
        {
          id: 'doc8',
          source: 'doc8.md',
          sourceDirName: 'subfolder/subsubfolder/subsubsubfolder3',
          sidebarPosition: 1,
          frontMatter: {},
        },
      ],
      options: {
        sidebarCollapsed: true,
        sidebarCollapsible: true,
      },
    });

    expect(sidebarSlice).toEqual([
      {
        type: 'category',
        label: 'subsubsubfolder3 (_category_.json label)',
        link: {
          id: 'doc1',
          type: 'doc',
        },
        items: [
          {type: 'doc', id: 'doc8'},
          {type: 'doc', id: 'doc7'},
        ],
      },
      {
        type: 'category',
        label: 'subsubsubfolder2 (_category_.yml label)',
        className: 'bar',
        items: [{type: 'doc', id: 'doc6'}],
      },
      {type: 'doc', id: 'doc1'},
      {type: 'doc', id: 'doc4'},
      {
        type: 'category',
        label: 'subsubsubfolder',
        items: [{type: 'doc', id: 'doc5'}],
      },
    ] as Sidebar);
  });

  test('uses explicit link over the index/readme.{md,mdx} naming convention', async () => {
    const sidebarSlice = await DefaultSidebarItemsGenerator({
      numberPrefixParser: DefaultNumberPrefixParser,
      item: {
        type: 'autogenerated',
        dirName: '.',
      },
      version: {
        versionName: 'current',
        contentPath: '',
      },
      categoriesMetadata: {
        Category: {
          label: 'Category label',
          link: {
            type: 'doc',
            id: 'doc3', // Using a "local doc id" ("doc1" instead of "parent/doc1") on purpose
          },
        },
        Category2: {
          label: 'Category 2 label',
          link: null,
        },
      },
      docs: [
        {
          id: 'parent/doc1',
          source: '@site/docs/Category/index.md',
          sourceDirName: 'Category',
          frontMatter: {},
        },
        {
          id: 'parent/doc2',
          source: '@site/docs/Category/index.md',
          sourceDirName: 'Category',
          frontMatter: {},
        },
        {
          id: 'parent/doc3',
          source: '@site/docs/Category/doc3.md',
          sourceDirName: 'Category',
          frontMatter: {},
        },
        {
          id: 'parent/doc4',
          source: '@site/docs/Category2/doc1.md',
          sourceDirName: 'Category2',
          frontMatter: {},
        },
        {
          id: 'parent/doc5',
          source: '@site/docs/Category2/index.md',
          sourceDirName: 'Category2',
          frontMatter: {},
        },
        {
          id: 'parent/doc6',
          source: '@site/docs/Category2/doc3.md',
          sourceDirName: 'Category2',
          frontMatter: {},
        },
      ],
      options: {
        sidebarCollapsed: true,
        sidebarCollapsible: true,
      },
    });

    expect(sidebarSlice).toEqual([
      {
        type: 'category',
        label: 'Category label',
        link: {
          id: 'parent/doc3',
          type: 'doc',
        },
        items: [
          {
            id: 'parent/doc1',
            type: 'doc',
          },
          {
            id: 'parent/doc2',
            type: 'doc',
          },
        ],
      },
      {
        type: 'category',
        label: 'Category 2 label',
        items: [
          {
            id: 'parent/doc4',
            type: 'doc',
          },
          {
            id: 'parent/doc5',
            type: 'doc',
          },
          {
            id: 'parent/doc6',
            type: 'doc',
          },
        ],
      },
    ] as Sidebar);
  });

  test('respects custom isCategoryIndex', async () => {
    const sidebarSlice = await DefaultSidebarItemsGenerator({
      numberPrefixParser: DefaultNumberPrefixParser,
      isCategoryIndex({fileName, directories}) {
        return (
          fileName.replace(
            `${DefaultNumberPrefixParser(
              directories[0],
            ).filename.toLowerCase()}-`,
            '',
          ) === 'index'
        );
      },
      item: {
        type: 'autogenerated',
        dirName: '.',
      },
      version: {
        versionName: 'current',
        contentPath: '',
      },
      categoriesMetadata: {},
      docs: [
        {
          id: 'intro',
          source: '@site/docs/intro.md',
          sourceDirName: '.',
          sidebarPosition: 1,
          frontMatter: {},
        },
        {
          id: 'tutorials-index',
          source: '@site/docs/01-Tutorials/tutorials-index.md',
          sourceDirName: '01-Tutorials',
          sidebarPosition: 2,
          frontMatter: {},
        },
        {
          id: 'tutorial2',
          source: '@site/docs/01-Tutorials/tutorial2.md',
          sourceDirName: '01-Tutorials',
          sidebarPosition: 2,
          frontMatter: {},
        },
        {
          id: 'tutorial1',
          source: '@site/docs/01-Tutorials/tutorial1.md',
          sourceDirName: '01-Tutorials',
          sidebarPosition: 1,
          frontMatter: {},
        },
        {
          id: 'not-guides-index',
          source: '@site/docs/02-Guides/README.md',
          sourceDirName: '02-Guides',
          frontMatter: {},
        },
        {
          id: 'guide2',
          source: '@site/docs/02-Guides/guide2.md',
          sourceDirName: '02-Guides',
          sidebarPosition: 2,
          frontMatter: {},
        },
        {
          id: 'guide1',
          source: '@site/docs/02-Guides/guide1.md',
          sourceDirName: '02-Guides',
          sidebarPosition: 1,
          frontMatter: {
            sidebar_class_name: 'foo',
          },
        },
      ],
      options: {
        sidebarCollapsed: true,
        sidebarCollapsible: true,
      },
    });

    expect(sidebarSlice).toEqual([
      {type: 'doc', id: 'intro'},
      {
        type: 'category',
        label: 'Tutorials',
        link: {
          type: 'doc',
          id: 'tutorials-index',
        },
        items: [
          {type: 'doc', id: 'tutorial1'},
          {type: 'doc', id: 'tutorial2'},
        ],
      },
      {
        type: 'category',
        label: 'Guides',
        items: [
          {type: 'doc', id: 'guide1', className: 'foo'},
          {type: 'doc', id: 'guide2'},
          {
            type: 'doc',
            id: 'not-guides-index',
          },
        ],
      },
    ] as Sidebar);
  });
});
