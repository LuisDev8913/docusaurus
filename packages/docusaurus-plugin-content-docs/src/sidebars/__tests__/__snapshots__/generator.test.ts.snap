// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultSidebarItemsGenerator generates complex nested sidebar 1`] = `
Array [
  Object {
    "id": "intro",
    "type": "doc",
  },
  Object {
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "id": "tutorial1",
        "type": "doc",
      },
      Object {
        "id": "tutorial2",
        "type": "doc",
      },
    ],
    "label": "Tutorials",
    "link": Object {
      "id": "tutorials-index",
      "type": "doc",
    },
    "type": "category",
  },
  Object {
    "collapsed": false,
    "collapsible": undefined,
    "customProps": Object {
      "description": "foo",
    },
    "items": Array [
      Object {
        "className": "foo",
        "id": "guide1",
        "type": "doc",
      },
      Object {
        "collapsed": undefined,
        "collapsible": undefined,
        "items": Array [
          Object {
            "id": "nested-guide",
            "type": "doc",
          },
        ],
        "label": "SubGuides (metadata file label)",
        "link": Object {
          "description": "subguides-description",
          "slug": "subguides-generated-index-slug",
          "title": "subguides-title",
          "type": "generated-index",
        },
        "type": "category",
      },
      Object {
        "id": "guide2",
        "type": "doc",
      },
    ],
    "label": "Guides",
    "link": Object {
      "id": "guides-index",
      "type": "doc",
    },
    "type": "category",
  },
  Object {
    "id": "end",
    "type": "doc",
  },
]
`;

exports[`DefaultSidebarItemsGenerator generates simple flat sidebar 1`] = `
Array [
  Object {
    "id": "doc3",
    "type": "doc",
  },
  Object {
    "id": "doc4",
    "type": "doc",
  },
  Object {
    "id": "doc1",
    "label": "doc1 sidebar label",
    "type": "doc",
  },
  Object {
    "id": "doc2",
    "type": "doc",
  },
  Object {
    "id": "doc5",
    "type": "doc",
  },
]
`;

exports[`DefaultSidebarItemsGenerator generates subfolder sidebar 1`] = `
Array [
  Object {
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "id": "doc8",
        "type": "doc",
      },
      Object {
        "id": "doc7",
        "type": "doc",
      },
    ],
    "label": "subsubsubfolder3 (_category_.json label)",
    "link": Object {
      "id": "doc1",
      "type": "doc",
    },
    "type": "category",
  },
  Object {
    "className": "bar",
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "id": "doc6",
        "type": "doc",
      },
    ],
    "label": "subsubsubfolder2 (_category_.yml label)",
    "type": "category",
  },
  Object {
    "id": "doc1",
    "type": "doc",
  },
  Object {
    "id": "doc4",
    "type": "doc",
  },
  Object {
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "id": "doc5",
        "type": "doc",
      },
    ],
    "label": "subsubsubfolder",
    "type": "category",
  },
]
`;

exports[`DefaultSidebarItemsGenerator respects custom isCategoryIndex 1`] = `
Array [
  Object {
    "id": "intro",
    "type": "doc",
  },
  Object {
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "id": "tutorial1",
        "type": "doc",
      },
      Object {
        "id": "tutorial2",
        "type": "doc",
      },
    ],
    "label": "Tutorials",
    "link": Object {
      "id": "tutorials-index",
      "type": "doc",
    },
    "type": "category",
  },
  Object {
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "className": "foo",
        "id": "guide1",
        "type": "doc",
      },
      Object {
        "id": "guide2",
        "type": "doc",
      },
      Object {
        "id": "not-guides-index",
        "type": "doc",
      },
    ],
    "label": "Guides",
    "type": "category",
  },
]
`;

exports[`DefaultSidebarItemsGenerator uses explicit link over the index/readme.{md,mdx} naming convention 1`] = `
Array [
  Object {
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "id": "parent/doc2",
        "type": "doc",
      },
      Object {
        "id": "parent/doc1",
        "type": "doc",
      },
    ],
    "label": "Category label",
    "link": Object {
      "id": "parent/doc3",
      "type": "doc",
    },
    "type": "category",
  },
  Object {
    "collapsed": undefined,
    "collapsible": undefined,
    "items": Array [
      Object {
        "id": "parent/doc4",
        "type": "doc",
      },
      Object {
        "id": "parent/doc6",
        "type": "doc",
      },
      Object {
        "id": "parent/doc5",
        "type": "doc",
      },
    ],
    "label": "Category 2 label",
    "type": "category",
  },
]
`;
